tp-json/tests/40_pokedex.cpp

```diff
@@ -0,0 +1,99 @@
+#include "custom_assert.hpp"
+
+#include <algorithm>
+#include <iostream>
+#include <unordered_map>
+#include <vector>
+
+int main(int argc, char** argv)
+{
+    if (argc < 2)
+    {
+        std::cout << "First command-line argument needs to be the path to the json resources.";
+        exit(EXIT_FAILURE);
+    }
+    std::string dir = std::string(argv[1]);
+    std::string filename;
+
+    {
+        Node_ptr node;
+
+        filename = dir + "pokedex.json";
+        std::cerr << "Starting test with: " << filename << std::endl;
+        node = JsonParser::parse_from_file(filename);
+
+        ASSERT_EQUAL(node->kind(), NodeKind::OBJECT)
+        ASSERT_UNEQUAL(node->as_ObjectNode()->at("pokemon"), nullptr)
+        ASSERT_EQUAL(node->as_ObjectNode()->at("pokemon")->kind(), NodeKind::ARRAY)
+
+        std::unordered_map<std::string, unsigned> pokemon_id_by_name;
+        for (const auto& pokemon_node : *(node->as_ObjectNode()->at("pokemon")->as_ArrayNode()))
+        {
+            ASSERT_EQUAL(pokemon_node->kind(), NodeKind::OBJECT)
+            auto pokemon_obj = pokemon_node->as_ObjectNode();
+
+            ASSERT_UNEQUAL(pokemon_obj->at("name"), nullptr)
+            ASSERT_EQUAL(pokemon_obj->at("name")->kind(), NodeKind::STRING)
+            const std::string& name = pokemon_obj->at("name")->as_StringLeaf()->data();
+
+            ASSERT_UNEQUAL(pokemon_obj->at("id"), nullptr)
+            ASSERT_UNEQUAL(pokemon_obj->at("id")->as_NumberLeaf(), nullptr)
+            unsigned id = pokemon_obj->at("id")->as_NumberLeaf()->data();
+
+            pokemon_id_by_name[name] = id;
+        }
+
+        std::unordered_map<std::string, std::vector<unsigned>> pokemon_id_by_type;
+        for (const auto& pokemon_node : *(node->as_ObjectNode()->at("pokemon")->as_ArrayNode()))
+        {
+            ASSERT_EQUAL(pokemon_node->kind(), NodeKind::OBJECT)
+            ASSERT_UNEQUAL(pokemon_node->as_ObjectNode()->at("type"), nullptr)
+            ASSERT_EQUAL(pokemon_node->as_ObjectNode()->at("type")->kind(), NodeKind::ARRAY)
+            for (const auto& type_node : *(pokemon_node->as_ObjectNode()->at("type")->as_ArrayNode()))
+            {
+                ASSERT_UNEQUAL(type_node->as_StringLeaf(), nullptr)
+                std::string const& type = type_node->as_StringLeaf()->data();
+                auto               it   = pokemon_id_by_type.find(type);
+                if (it == pokemon_id_by_type.end())
+                    it = pokemon_id_by_type.emplace(type, std::vector<unsigned>()).first;
+                it->second.emplace_back(pokemon_node->as_ObjectNode()->at("id")->as_NumberLeaf()->data());
+            }
+        }
+
+        for (const auto& pair : pokemon_id_by_type)
+        {
+            std::cerr << pair.first << ": [";
+            bool b = false;
+            for (auto i : pair.second)
+            {
+                if (b)
+                    std::cerr << ", ";
+                else
+                    b = true;
```

Comment:
https://stackoverflow.com/questions/10750057/how-do-i-print-out-the-contents-of-a-vector/45400359#45400359

Reply:
Je ne suis pas fan des *one-liner* en général.

Reply:
c'est pas pour avoir un 1-liner, c'est pour ne pas avoir un JMP dans chaque passage du for :P

Reply:
Ce type d'optimisation me semble disproportionné pour un affichage dans un test.  Je préfère favoriser la lisibilité.

Par ailleurs, je pense que ce n'est pas évident que ce code soit plus efficace.
- C'est le test qui coûte cher, pas le jump, et on fait déjà un test pour le tour de boucle.
- Ce genre de test est facile à prédire par le processeur à l'execution (il est écrit pour être égal à true la plupart du temps)
- On rajoute un coût constant pour construire un objet en plus et faire des appels de fonctions (et on instancie un template ce qui augmente la taille du code)
- On ne sait pas ce que fait le ostream_operator, qui pourrait faire lui-même des tests de sécurité (j'ai été assez déçu de l'efficacité des ostream/istream par rapports aux fonctions de bases en C).
Suivant la taille typique du vecteur concerné, peut-être que cette solution n'est pas plus rapide.   Il faudrait tester.

Reply:
il apparait que le gcc est assez bien pour enlever la variable et juste doubler l'appel à `cout << ` pour le premier element, du coup, ça me va :)